yaptide/routes/keycloak_routes.py

```diff
@@ -13,10 +15,50 @@
 from yaptide.routes.utils.response_templates import (error_internal_response,
                                                      yaptide_response)
 from yaptide.routes.utils.tokens import encode_auth_token
+from werkzeug.exceptions import Forbidden, Unauthorized
 
 ROOT_DIR = Path(__file__).parent.resolve()
 
 
+def check_user_based_on_keycloak_token(token: str, username: str) -> str:
+    """Checks if user can access the service"""
+    if not token:
+        logging.error("No token provided")
+        raise Unauthorized(description="No token provided")
+    keycloak_base_url = os.environ.get('KEYCLOAK_BASE_URL', '')
+    keycloak_realm = os.environ.get('KEYCLOAK_REALM', '')
+    if not keycloak_base_url or not keycloak_realm:
+        logging.error("Keycloak env variables not set")
+        raise Forbidden(description="Service is not available")
+    keycloak_url = f"{keycloak_base_url}/auth/realms/{keycloak_realm}/protocol/openid-connect/certs"
+    try:
+        unverified_encoded_token = jwt.decode(token, options={"verify_signature": False})
+        if username != unverified_encoded_token["preferred_username"]:
+            logging.error("Username mismatch")
+            raise Forbidden(description="Username mismatch")
+        res = requests.get(keycloak_url)
+        jwks = res.json()
+
+        public_keys = {}
+        for jwk in jwks['keys']:
+            kid = jwk['kid']
+            public_keys[kid] = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(jwk))
+
+        kid = jwt.get_unverified_header(token)['kid']
+        key = public_keys[kid]
+
+        _ = jwt.decode(token, key=key, audience=unverified_encoded_token["aud"], algorithms=['RS256'])
```

Comment:
```suggestion
        _ = jwt.decode(token, key=key, audience=unverified_encoded_token["aud"], algorithms=['RS256'], options={"verify_signature": True})
```

Change suggested by bsobol